src/components/AlbumList.tsx

```diff
@@ -0,0 +1,129 @@
+import * as React from 'react';
+import { Album } from '../interfaces/album';
+import { Song } from '../interfaces/song';
+
+import { AlbumListItem } from './AlbumListItem';
+
+interface Props {
+  albums: Array<Album>;
+  searchQuery: string;
+  songs: Array<Song>;
+  onSongClick: Function;
+}
+
+interface AlbumHashLookup {
+  [key: number]: Album;
+}
+
+interface SongHashLookup {
+  [key: number]: Array<Song>;
+}
+
+export class AlbumList extends React.Component<Props, undefined> {
+  props: Props;
+
+  private sortedAlbums: Array<Album>;
+  private albumLookup: AlbumHashLookup;
+  private songAlbumLookup: SongHashLookup;
+
+  private albumMatches: Array<Album>;
+  private songMatches: SongHashLookup;
+  private indirectAlbumMatches: Set<number>;
+
+  constructor(props: Props) {
+    super(props);
+
+    this.sortAlbums();
+    this.optimizeData();
+
+    this.albumMatches = [];
+    this.songMatches = {};
+    this.indirectAlbumMatches = new Set<number>();
+  }
+
+  componentWillReceiveProps(nextProps: Props) {
+    if (nextProps.searchQuery !== this.props.searchQuery) {
+      this.performSearchQuery(nextProps.searchQuery);
+    }
+  }
+
+  render() {
+    const isSearching = this.props.searchQuery !== '';
+    const albumSource = isSearching ? this.albumMatches : this.sortedAlbums;
+    return (
+      <ul className="albumList">
+        {albumSource.map(album => {
+          const isDirectMatch = !this.indirectAlbumMatches.has(album.id);
+          return (
+            <AlbumListItem 
+              album={album} 
+              directSongMatches={this.songMatches[album.id]}
+              key={`album-list-${album.id}`}
+              songs={isSearching && !isDirectMatch
+                ? this.songMatches[album.id]
+                : this.songAlbumLookup[album.id]
+              } 
+              onSongClick={this.props.onSongClick}
+              isDirectMatch={isDirectMatch}
+              isSearching={isSearching}
+            />
+          )
+        })}
+      </ul>
+    );
+  }
+
+  sortAlbums() {
+    this.sortedAlbums = this.props.albums

```

Comment:
sort mutates the original array: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort